buildkite/resource_pipeline.go

```diff
@@ -717,22 +717,18 @@ func (p *pipelineResource) UpgradeState(ctx context.Context) map[int64]resource.
 }
 
 func (p *pipelineResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
-	// Only modify plan on pipeline creation
-	if req.State.Raw.IsNull() {
+	// Don't modify on destroy, but otherwise make sure the steps default is preserved
+	if !req.Plan.Raw.IsNull() {
 		var template, steps types.String
 
 		// Load pipeline_template_id and steps (if defined)
 		req.Plan.GetAttribute(ctx, path.Root("pipeline_template_id"), &template)
 		req.Plan.GetAttribute(ctx, path.Root("steps"), &steps)
 
-		// Set default steps only if there is no template oe defined steps
+		// Set default steps only if there is no template or defined steps
 		if template.IsNull() && (steps.IsUnknown() || steps.IsNull()) {
 			resp.Diagnostics.Append(resp.Plan.SetAttribute(ctx, path.Root("steps"), defaultSteps)...)
```

Comment:
```suggestion
if template.IsNull() && steps.IsNull() {
			resp.Diagnostics.Append(resp.Plan.SetAttribute(ctx, path.Root("steps"), defaultSteps)...)
```

Wondering if it makes sense to drop `steps.IsUnknown()` from this condition, to allow for when the value of `steps` is still be be determined, as we've got a case where an user has the value of `steps` be created elsewhere and is subject to change. As currently, it would set the 'default steps' in-place and result in a difference in the planned value and error out.

Reply:
Is this off the back of this [doc](https://github.com/hashicorp/terraform/blob/main/docs/resource-instance-change-lifecycle.md) specifying how this is called twice? I think you might be on the right track, although I'm wondering if the "unknown" check should be moved way up to the top - something like, if its unknown, do nothing straight away - and wait for the invocation where it is known

Reply:
I completely removed the checks that were intended for "creating/updating" as it doesn't really matter.

At the end of it all, we care what the user has supplied (req.Config) and whether we are in "template" or "steps" mode (explicit or implicit).

I've updated the plan modifier and created a few new tests for each case. I think it covers it all. And they are all passing!! Have a look over and see if there are any branches I haven't covered in the tests